src/optparse.coffee

```diff
@@ -75,41 +67,98 @@ exports.OptionParser = class OptionParser
-buildRules = (rules) ->
-  for tuple in rules
+buildRules = (ruleDecls) ->
+  ruleList = for tuple in ruleDecls
     tuple.unshift null if tuple.length < 3
     buildRule tuple...
+  flagDict = {}
+  for rule in ruleList
+    # shortFlag is null if not provided in the rule.
+    for flag in [rule.shortFlag, rule.longFlag] when flag?
+      if flagDict[flag]?
+        throw new Error "flag #{flag} for switch #{rule.name}
+          was already declared for switch #{flagDict[flag].name}"
+      flagDict[flag] = rule
+
+  {ruleList, flagDict}
 
 # Build a rule from a `-o` short flag, a `--output [DIR]` long flag, and the
 # description of what the option does.
-buildRule = (shortFlag, longFlag, description, options = {}) ->
+buildRule = (shortFlag, longFlag, description) ->
   match     = longFlag.match(OPTIONAL)
+  shortFlag = shortFlag?.match(SHORT_FLAG)[1]
   longFlag  = longFlag.match(LONG_FLAG)[1]
   {
-    name:         longFlag.substr 2
+    name:         longFlag.replace /^--/, ''
     shortFlag:    shortFlag
     longFlag:     longFlag
     description:  description
     hasArgument:  !!(match and match[1])
     isList:       !!(match and match[2])
   }
 
-# Normalize arguments by expanding merged flags into multiple flags. This allows
-# you to have `-wl` be the same as `--watch --lint`.
-normalizeArguments = (args) ->
-  args = args[..]
-  result = []
-  for arg in args
-    if match = arg.match MULTI_FLAG
-      result.push '-' + l for l in match[1].split ''
+normalizeArguments = (args, flagDict) ->
+  rules = []
+  positional = []
+  needsArgOpt = null
+  for arg, argIndex in args
+    # If the previous argument given to the script was an option that uses the
+    # next command-line argument as its argument, create copy of the option's
+    # rule with an `argument` field.
+    if needsArgOpt?
+      # FIXME: use object spread when that gets merged (in #4493)
+      rules.push Object.assign({}, needsArgOpt.rule, {argument: arg})
+      needsArgOpt = null
+      continue
+
+    multiFlags = arg.match(MULTI_FLAG)?[1]
+      .split('')
+      .map (flagName) -> "-#{flagName}"
+    if multiFlags?
+      multiOpts = multiFlags.map (flag) ->
+        rule = flagDict[flag]
+        unless rule?
+          throw new Error "unrecognized option #{flag} in multi-flag #{arg}"
+        {rule, flag}
+      # Only the last flag in a multi-flag may have an argument.
+      [innerOpts..., lastOpt] = multiOpts
+      for {rule, flag} in innerOpts
+        if rule.hasArgument
+          throw new Error "cannot use option #{flag} in multi-flag #{arg} except
+          as the last option, because it needs an argument"
+        rules.push rule
+      if lastOpt.rule.hasArgument
+        needsArgOpt = lastOpt
+      else
+        rules.push lastOpt.rule
+    else if ([LONG_FLAG, SHORT_FLAG].some (pat) -> arg.match(pat)?)
+      # TODO: do we need to check if `arg` matches regex if we already check if
+      # it's in flagDict?
```

Comment:
Checking for multi-flags makes sense, since they have to be split into individual flags that we do immediately recognize. I was thinking it was a little redundant to check if an argument matches a regex when we do exact string matching the line right after, in `flagDict[arg]`. This comment can be removed, however, because we want to notify the user if they put in an option that doesn't exist.